### PR TITLE
De-duplicated convolutional layer code.

### PR DESCRIPTION
**Pull request by MikeAmy**
Extracted AbstractConvolution super class.
Convolution layers have a common 'dimensions' attribute that gives the dimensions.

### CODE REVIEW

**Comment by MikeAmy** [comment]
Just pushing here now to await Travis as I'm too lazy to test on both backends on my machines.

Incidentally, why does Convolution1D differ from Convolution2D and 3D in that it doesn't care about Tensorflow/Theano dim orderings? I know it's 1D but the input shape is handled differently and still looks like it has more than one dimension. (I guess I should just play with it more).

**Comment by MikeAmy** [comment]
By the way this removes 80 something lines. I wondered if we could go further and create a Convolution layer that handles any number of dimensions. That could be useful.

File: keras/layers/convolutional.py [hunk 1]
```diff
@@ -11,7 +11,90 @@
...
+                    init=self.init.__name__,
+                    activation=self.activation.__name__,
+                    border_mode=self.border_mode,
+                    W_regularizer=get_config_if_not_none(self.W_regularizer),
+                    b_regularizer=get_config_if_not_none(self.b_regularizer),
+                    activity_regularizer=get_config_if_not_none(self.activity_regularizer),
+                    W_constraint=get_config_if_not_none(self.W_constraint),
+                    b_constraint=get_config_if_not_none(self.b_constraint),
+                    bias=self.bias,
+                    **child_class_config)
```
Comment (MikeAmy) [inline keras/layers/convolutional.py]: btw, the nice things about constructing dicts in this way is that the child class just calls super with the extra bits, and because we are using keyword arguments, python does a check on the arguments that none are duplicated, which catches certain mistakes. Of course, if a child class wanted to override an argument, it can still do so, but it will be explicit, which helps readability. Picked this up from working working with django.

File: keras/layers/convolutional.py [hunk 2]
```diff
@@ -256,70 +286,26 @@ def __init__(self, nb_filter, nb_row, nb_col,
...
             stack_size = input_shape[1]
-            self.W_shape = (self.nb_filter, stack_size, self.nb_row, self.nb_col)
+            self.W_shape = (self.nb_filter, stack_size) + self.dimensions
         elif self.dim_ordering == 'tf':
             stack_size = input_shape[3]
-            self.W_shape = (self.nb_row, self.nb_col, stack_size, self.nb_filter)
+            self.W_shape = self.dimensions + (stack_size, self.nb_filter)
         else:
             raise Exception('Invalid dim_ordering: ' + self.dim_ordering)
-        self.W = self.init(self.W_shape, name='{}_W'.format(self.name))
```
Comment (MikeAmy) [inline keras/layers/convolutional.py]: This code is duplicated pretty much verbatim three times. Moved it up.

File: keras/layers/convolutional.py [hunk 3]
```diff
@@ -434,83 +407,38 @@ class Convolution3D(Layer):
...
     def __init__(self, nb_filter, kernel_dim1, kernel_dim2, kernel_dim3,
                  init='glorot_uniform', activation='linear', weights=None,
                  border_mode='valid', subsample=(1, 1, 1), dim_ordering=K.image_dim_ordering(),
                  W_regularizer=None, b_regularizer=None, activity_regularizer=None,
                  W_constraint=None, b_constraint=None,
                  bias=True, **kwargs):
-        if border_mode not in {'valid', 'same'}:
-            raise Exception('Invalid border mode for Convolution3D:', border_mode)
-        self.nb_filter = nb_filter
+        super(Convolution3D, self).__init__(nb_filter,
```
Comment (MikeAmy) [inline keras/layers/convolutional.py]: Working hard to follow the indentation style but to be honest it's pretty tiring having to keep on tabbing and spacing things into position, and the lines easily run over 80 chars. I often work with multiple IDE windows side by side so 80 char limits help to avoid word-wrap/having to scroll.

How strict is this style? Is it a problem if some code is indented e.g.:

        super(Convolution3D, self).__init__(
            nb_filter,
            dimensions=(kernel_dim1, kernel_dim2, kernel_dim3),
            init=init,
            ...
            bias=bias,
            **kwargs
        )

File: keras/layers/convolutional.py [hunk 4]
```diff
@@ -256,70 +286,26 @@ def __init__(self, nb_filter, nb_row, nb_col,
...
         self.nb_col = nb_col
-        self.init = initializations.get(init, dim_ordering=dim_ordering)
-        self.activation = activations.get(activation)
-        assert border_mode in {'valid', 'same'}, 'border_mode must be in {valid, same}'
-        self.border_mode = border_mode
-        self.subsample = tuple(subsample)
-        assert dim_ordering in {'tf', 'th'}, 'dim_ordering must be in {tf, th}'
-        self.dim_ordering = dim_ordering
-
-        self.W_regularizer = regularizers.get(W_regularizer)
```
Comment (MikeAmy) [inline keras/layers/convolutional.py]: These lines are duplicated pretty much verbatim.

File: keras/layers/convolutional.py (hunk 3 above)
Comment (fchollet) [inline keras/layers/convolutional.py]: No, that's a fine style as well. Anything that's PEP8 works. And line length is not strictly enforced (prefer readability over correct line length).

**Comment by fchollet** [comment]
I think an `AbstractConvolution` class is much needed. We are about to add several new convolution layers (`AtrousConv2D`, `SeparableConv2D`, `Deconv2D`) and I think we should wait until we have them around, since they would also be inheriting from `AbstractConvolution`. It would be too bad to have to modify the abstraction multiple times to accommodate new layers. What do you think?

File: keras/layers/convolutional.py (hunk 3 above)
Comment (MikeAmy) [inline keras/layers/convolutional.py]: Happy to hear. :)

**Comment by MikeAmy** [comment]
Yes, totally understand, in fact I already had a close look at the atrousConv2d PR for this exact reason, so I ended up reviewing the PR. I will be more than happy to look at the code for those other layers and help work them in if needed.

Let me look again at that Atrous2D code. My suspicion is that it's not an issue as remember the AbstractConvolution class is not published API at the moment, so only 'fallout' is PR conflicts, but I get your point. I think Deconv2D may be an altogether different beast, SeparableConv2D and AtrousConv2D won't be difficult. Maybe I'm being optimistic though.

**Comment by MikeAmy** [comment]
I checked the additions to Convolution2D in the Atrous convolution PR:

https://github.com/fchollet/keras/pull/3183

None of them would affect the AbstractConvolution class, they are all in the Convolution2D class.

**Comment by MikeAmy** [comment]
If the Deconv2D is the same as this one:

http://caffe.berkeleyvision.org/doxygen/classcaffe_1_1DeconvolutionLayer.html#details

Then I think that the Deconv2D class looks like it will be exactly the same as the Convolution2D class except that the backend function will differ.

**Comment by MikeAmy** [comment]
Is there any code for Deconv2D and SeparableConv2D that I can look at?

**Comment by fchollet** [comment]
Not yet, but @EderSantana was working on it, so maybe check with him!

**Comment by EderSantana** [comment]
Right now I'm using it like this. But I want to write a backend independent version. This one is TF only
```python
class Deconv2D(Layer):
    def __init__(self, nb_filter, nb_row, nb_col,
                 init='glorot_uniform', activation='linear', weights=None,
                 border_mode='valid', subsample=(1, 1), dim_ordering='tf',
                 W_regularizer=None, b_regularizer=None, activity_regularizer=None,
                 W_constraint=None, b_constraint=None, **kwargs):

        if border_mode not in {'valid', 'same'}:
            raise Exception('Invalid border mode for Convolution2D:', border_mode)
        self.nb_filter = nb_filter
        self.nb_row = nb_row
        self.nb_col = nb_col
        self.init = initializations.get(init, dim_ordering=dim_ordering)
        self.activation = activations.get(activation)
        assert border_mode in {'valid', 'same'}, 'border_mode must be in {valid, same}'
        self.border_mode = border_mode
        self.subsample = tuple(subsample)
        assert dim_ordering in {'tf', 'th'}, 'dim_ordering must be in {tf, th}'
        self.dim_ordering = dim_ordering

        self.W_regularizer = regularizers.get(W_regularizer)
        self.b_regularizer = regularizers.get(b_regularizer)
        self.activity_regularizer = regularizers.get(activity_regularizer)

        self.W_constraint = constraints.get(W_constraint)
        self.b_constraint = constraints.get(b_constraint)

        self.input_spec = [InputSpec(ndim=4)]
        self.initial_weights = weights
        super(Deconv2D, self).__init__(**kwargs)

    def build(self, input_shape):
        if self.dim_ordering == 'th':
            stack_size = input_shape[1]
            self.W_shape = (self.nb_filter, stack_size, self.nb_row, self.nb_col)
        elif self.dim_ordering == 'tf':
            stack_size = input_shape[3]
            self.W_shape = (self.nb_row, self.nb_col, self.nb_filter, stack_size)
        else:
            raise Exception('Invalid dim_ordering: ' + self.dim_ordering)
        self.W = self.init(self.W_shape, name='{}/w'.format(self.name))
        self.b = K.zeros((self.nb_filter,), name='{}/biases'.format(self.name))
        self.trainable_weights = [self.W, self.b]
        self.regularizers = []

        if self.W_regularizer:
            self.W_regularizer.set_param(self.W)
            self.regularizers.append(self.W_regularizer)

        if self.b_regularizer:
            self.b_regularizer.set_param(self.b)
            self.regularizers.append(self.b_regularizer)

        if self.activity_regularizer:
            self.activity_regularizer.set_layer(self)
            self.regularizers.append(self.activity_regularizer)

        self.constraints = {}
        if self.W_constraint:
            self.constraints[self.W] = self.W_constraint
        if self.b_constraint:
            self.constraints[self.b] = self.b_constraint

        if self.initial_weights is not None:
            self.set_weights(self.initial_weights)
            del self.initial_weights

    def get_output_shape_for(self, input_shape):
        if self.dim_ordering == 'th':
            rows = input_shape[2]
            cols = input_shape[3]
        elif self.dim_ordering == 'tf':
            rows = input_shape[1]
            cols = input_shape[2]
        else:
            raise Exception('Invalid dim_ordering: ' + self.dim_ordering)

        rows = rows * self.subsample[0]
        cols = cols * self.subsample[1]

        if self.dim_ordering == 'th':
            return (input_shape[0], self.nb_filter, rows, cols)
        elif self.dim_ordering == 'tf':
            return (input_shape[0], rows, cols, self.nb_filter)
        else:
            raise Exception('Invalid dim_ordering: ' + self.dim_ordering)

    def call(self, x, mask=None):
        output_shape = self.get_output_shape_for(x.get_shape().as_list())
        deconv_out = tf.nn.conv2d_transpose(
            x, self.W, output_shape=output_shape, strides=[1, self.subsample[0], self.subsample[1], 1])

        if self.dim_ordering == 'th':
            output = deconv_out + K.reshape(self.b, (1, self.nb_filter, 1, 1))
        elif self.dim_ordering == 'tf':
            output = deconv_out + K.reshape(self.b, (1, 1, 1, self.nb_filter))
        else:
            raise Exception('Invalid dim_ordering: ' + self.dim_ordering)
        output = self.activation(output)
        return output

    def get_config(self):
        config = {'nb_filter': self.nb_filter,
                  'nb_row': self.nb_row,
                  'nb_col': self.nb_col,
                  'init': self.init.__name__,
                  'activation': self.activation.__name__,
                  'border_mode': self.border_mode,
                  'subsample': self.subsample,
                  'dim_ordering': self.dim_ordering,
                  'W_regularizer': self.W_regularizer.get_config() if self.W_regularizer else None,
                  'b_regularizer': self.b_regularizer.get_config() if self.b_regularizer else None,
                  'activity_regularizer': self.activity_regularizer.get_config() if self.activity_regularizer else None,
                  'W_constraint': self.W_constraint.get_config() if self.W_constraint else None,
                  'b_constraint': self.b_constraint.get_config() if self.b_constraint else None}
        base_config = super(Deconv2D, self).get_config()
        return dict(list(base_config.items()) + list(config.items()))
```

**Comment by MikeAmy** [comment]
Thanks a lot. I've started looking at it.

**Comment by fchollet** [comment]
All additional conv layers have now been merged (`Deconv2D`, `AtrousConv2D`, SeparableConv2D`). Several of them are using `Conv2D` as their parent class. If you are still interested, you can look at how to make the current setup more concise and elegant!

**Comment by MikeAmy** [comment]
Excellent, will have a look. Sorry, I've been traveling in Japan and UK lately, so a bit slow to respond.

**Comment by fchollet** [comment]
Closing because the PR is outdated. In the future, feel free to reopen a PR to abstract common parts of `Conv2D`, `AtrousConv2D`, etc.

**Comment by MikeAmy** [comment]
Sure thanks. I will.

Didn't because right now I am building a system to feed over 1 million
syrian refugees in Lebanon for UN. Pretty good excuse huh? :)
